liesel/band.py

```diff
@@ -0,0 +1,232 @@
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    a = to_utri_full(band, lower)
+    return a + jnp.tril(a.T, k=-1)
+
+
+@jax.jit
+def to_ltri_full(band, lower=True):
+    """
+    Returns the lower triangular full (n x n) matrix represented by `band`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    return to_utri_full(band, lower).T
+
+
+@jax.jit
+def to_utri_full(band, lower=True):
+    """
+    Returns the upper triangular full (n x n) matrix represented by `band`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band with bandwidth p+1.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+    p, n = band.shape
+    p -= 1
+    return jax.lax.cond(lower,
+                        lambda: sum([jnp.diag(v[:n - k], k) for k, v in enumerate(band)]),  # lower
+                        lambda: sum([jnp.diag(band[k][p - k:], p - k) for k in range(p + 1)]))  # upper
+
+
+@jax.jit
+def to_lower(band):
+    """
+    ## Parameters
+
+    - `band`: A (p x n) band in upper-diagonal ordered form.
+    """
+    return jnp.flipud(jnp.array([jnp.roll(row, i - band.shape[0] + 1) for i, row in enumerate(band)]))
+
+
+@jax.jit
+def to_upper(band):
+    """
+    ## Parameters
+
+    - `band`: A (p x n) band in lower-diagonal ordered form.
+    """
+    return jnp.flipud(jnp.array([jnp.roll(row, i) for i, row in enumerate(band)]))
+
+
+@jax.jit
+def cholesky(band, lower=True):
+    """
+    Returns the Cholesky decomposition of `band` as a (p x n) band in lower-diagonal
+    ordered form.
+
+    ## Parameters
+
+    - `band`: A (p x n) band.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+
+    band = jax.lax.cond(lower, lambda x: x, to_lower, band)
+
+    p = band.shape[0] - 1
+    n = band.shape[1]
+
+    L = jnp.zeros_like(band)
+
+    def outer(j, L):
+        v = band[:, j]
+
+        def inner(k, v):
+            def add(i, v):
+                return v.at[i - j + k].add(-(L[i, k] * L[j - k, k]))
+
+            return jax.lax.fori_loop(j - k, p + 1, add, v)
+
+        start = jnp.clip(j - p, a_min=0, a_max=None)
+        v = jax.lax.fori_loop(start, j, inner, v)
+
+        return L.at[:, j].set(v / jnp.sqrt(v[0]))
+
+    return jax.lax.fori_loop(0, n, outer, L)
+
+
+@jax.jit
+def forward_sub(band, rhs, lower=True):
+    """
+    Solves `L @ x = b`, where `L = to_ltri_full(band, lower)` and `b = rhs`.
+
+    ## Parameters
+
+    - `band`: A (p x n) band.
+    - `rhs`: The right-hand side of the system of linear equations.
+    - `lower`: Whether `band` is in lower-diagonal ordered form.
+    """
+
+    band = jax.lax.cond(lower, to_upper, lambda x: x, band)
+
+    p = band.shape[0] - 1
+    n = band.shape[1]
+
+    x = jnp.zeros(n + p)
+
+    for i in range(n):
```

Comment:
Can we use a `fori_loop` here now?

Reply:
Yes, implemented with `jax.lax.dynamic_slice`.